ranking/src/ts_system/mod.rs

```diff
@@ -0,0 +1,314 @@
+
+use rayon::prelude::*;
+use std::cell::{RefCell, RefMut};
+use std::cmp::max;
+use std::collections::{HashMap, VecDeque};
+use std::collections::hash_map::Entry;
+
+use std::rc::{Rc, Weak};
+use std::f64::INFINITY;
+
+type TSPlayerRating = Gaussian;
+type TSMessage = nodes::Message;
+type TSPlayer = String;
+type TSTeam = Vec<TSPlayer>;
+type TSContestPlace = Vec<TSTeam>;
+type TSContest = Vec<TSContestPlace>;
+type TSRating = HashMap<TSPlayer, TSPlayerRating>;
+
+// TrueSkillStPB rating system
+pub struct TrueSkillSPBSystem {
+    // epsilon used for ties
+    eps: f64,
+    // default player rating
+    mu: f64,
+    // default player sigma
+    sigma: f64,
+    // performance sigma
+    beta: f64,
+    // epsilon used for convergence loop
+    convergence_eps: f64,
+    // defines sigma growth per second
+    sigma_growth: f64,
+    // history of ratings for all players
+    rating: TSRating,
+}
+
+impl Default for TrueSkillSPBSystem {
+    fn default() -> Self {
+        Self {
+            eps: 0.70,
+            mu: 1500.,
+            sigma: 1500. / 3., // mu/3
+            beta: 1500. / 6., // sigma/2
+            convergence_eps: 2e-4,
+            sigma_growth: 5.,
+            rating: TSRating::new(),
+        }
+    }
+}
+
+// fn update_rating(old: &TSRating, new: &mut TSRatingHistory, contest: &TSContest, when: usize) {
+//     for place in &contest[..] {
+//         for team in &place[..] {
+//             for player in &team[..] {
+//                 new.entry(player.clone()).or_insert(Vec::new()).push((old.get(player).unwrap().clone(), when));
+//             }
+//         }
+//     }
+// }
+
+
+fn gen_team_message<T, K: Clone>(places: &Vec<Vec<T>>, default: &K) -> Vec<Vec<K>> {
+    let mut ret: Vec<Vec<K>> = Vec::with_capacity(places.len());
+
+    for place in places {
+        ret.push(vec![default.clone(); place.len()]);
+    }
+
+    ret
+}
+
+
+fn gen_player_message<T, K: Clone>(places: &Vec<Vec<Vec<T>>>, default: &K) -> Vec<Vec<Vec<K>>> {
+    let mut ret = Vec::with_capacity(places.len());
+
+    for place in places {
+        ret.push(Vec::with_capacity(place.len()));
+
+        for team in place {
+            ret.last_mut().unwrap().push(vec![default.clone(); team.len()]);
+        }
+    }
+
+    ret
+}
+
+
+fn infer1(who: &mut Vec<impl TreeNode>) {
+    for item in who {
+        item.infer();
+    }
+}
+
+
+fn infer2(who: &mut Vec<Vec<impl TreeNode>>) {
+    for item in who {
+        infer1(item);
+    }
+}
+
+
+fn infer3(who: &mut Vec<Vec<Vec<impl TreeNode>>>) {
+    for item in who {
+        infer2(item);
+    }
+}
+
+
+fn infer_ld(ld: &mut Vec<impl TreeNode>, l: &mut Vec<impl TreeNode>) {
+    for i in 0..ld.len() {
+        l[i].infer();
+        ld[i].infer();
+    }
+    l.last_mut().unwrap().infer();
+    for j in 0..ld.len() {
+        let i = ld.len() - 1 - j;
+        ld[i].infer();
+        l[i].infer();
+    }
+}
+
+
+fn check_convergence(a: &Vec<Rc<RefCell<(TSMessage, TSMessage)>>>,

```

Comment:
Please note that `&[T]` is a strictly more general argument type than `&Vec<T>`. Hence, the latter is uncommon.
Likewise for `&str` vs `&String`.